eclipse-theia/theia
# Notebook API Support

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/8395.
Future Improvements: https://github.com/eclipse-theia/theia/issues/12863.

This implements the basics of the notebook api for theia. This includes following features:

- opening notebooks in a custom editor widget
- editing and saving of notebooks through the new widget
- executing cells and notebooks and displaying cell outputs through renderers contributed by extensions
- general starting and comunicating with notebook extensions like jupyter

Also this PR includes some prepared functionality not in use yet but needed for implementations later on.

To keep the size of the PR on the smaller size (it's already gigantic as it is), some less important notebook features haven't been implemented yet. Don't expect everything to work perfectly as expected
see https://github.com/eclipse-theia/theia/pull/12442#issuecomment-1670782244 for information on known issues and missing features.

- [x] [IP Check Ticket](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/10085#note_1198130)

#### How to test
Basic Notebook support is given through the builtin ipynb extension. So execute `yarn download:plugins` once in the project root.
For Cell execution the [python](https://open-vsx.org/extension/ms-python/python) and [juypter](https://open-vsx.org/extension/ms-toolsai/jupyter) extensions are needed.

Make sure that the workspace you're using for testing is trusted or you have disabled the *workspace trust* feature. Otherwise, some features, such as notebook kernel selection might not work as expected.

Also here is a notebook for testing  [W39Mon_Lecture8_DemoNotebook.zip](https://github.com/eclipse-theia/theia/files/11296571/W39Mon_Lecture8_DemoNotebook.zip)

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
